File: src/main/java/gargoyle/xenox/util/log/Log.java
package gargoyle.xenox.util.log;

import java.util.Date;

@SuppressWarnings("LawOfDemeter")
public final class Log {
    private static final Logger[] LOGGERS = {
            new JULLogger(null)
    };

    private Log() {
    }

    public static void debug(String message) {
        log(message, LEVEL.DEBUG);
    }

    public static void error(String message) {
        log(message, LEVEL.ERROR);
    }

    public static void error(String message, Throwable throwable) {
        log(message, throwable, LEVEL.ERROR);
    }

    public static void fatal(String message) {
        log(message, LEVEL.FATAL);
    }

    public static void fatal(String message, Throwable throwable) {
        log(message, throwable, LEVEL.FATAL);
    }

    private static Logger findLogger(LEVEL level) {
        for (Logger logger : LOGGERS) {
            if (logger.canLog(level)) {
                return logger;
            }
        }
        return SystemLogger.getSystemLogger();
    }

    private static StackTraceElement getCaller() {
        StackTraceElement[] stackTrace = Thread.currentThread().getStackTrace();
        String name = Log.class.getPackage().getName();
        for (int i = 1, stackTraceLength = stackTrace.length; i < stackTraceLength; i++) {
            StackTraceElement element = stackTrace[i];
            String className = element.getClassName();
            if (!className.startsWith(name)) {
                return element;
            }
        }
        return stackTrace[0];
    }

    public static void info(String message) {
        log(message, LEVEL.INFO);
    }

    private static void log(String message, LEVEL level) {
        findLogger(level).log(getCaller(), new Date().getTime(), level, message, null);
    }

    public static void log(String message, Throwable throwable) {
        log(message, throwable, LEVEL.FATAL);
    }

    private static void log(String message, Throwable throwable, LEVEL level) {
        Logger logger = findLogger(level);
        logger.log(getCaller(), new Date().getTime(), level, message == null ? String.format("%s: %s", throwable.getClass(), throwable.getLocalizedMessage()) : message, throwable);
    }

    public static void warn(String message) {
        log(message, LEVEL.WARN);
    }

    public static void warn(String msg, Throwable throwable) {
        log(msg, throwable, LEVEL.WARN);
    }
}

File: src/main/kotlin/gargoyle/xenox/util/log/Log.kt
package gargoyle.xenox.util.log

import java.util.Date

object Log {
    private val LOGGERS = arrayOf<Logger>(
        JULLogger(LEVEL.DEBUG)
    )

    fun debug(message: String) {
        log(message, LEVEL.DEBUG)
    }

    fun error(message: String) {
        log(message, LEVEL.ERROR)
    }

    fun error(message: String?, throwable: Throwable) {
        log(message, throwable, LEVEL.ERROR)
    }

    fun fatal(message: String) {
        log(message, LEVEL.FATAL)
    }

    fun fatal(message: String?, throwable: Throwable) {
        log(message, throwable, LEVEL.FATAL)
    }

    private fun findLogger(level: LEVEL): Logger {
        for (logger in LOGGERS) {
            if (logger.canLog(level)) {
                return logger
            }
        }
        return SystemLogger.systemLogger
    }

    private val caller: StackTraceElement
        get() {
            val stackTrace = Thread.currentThread().stackTrace
            val name = Log::class.java.getPackage().name
            var i = 1
            val stackTraceLength = stackTrace.size
            while (i < stackTraceLength) {
                val element = stackTrace[i]
                val className = element.className
                if (!className.startsWith(name)) {
                    return element
                }
                i++
            }
            return stackTrace[0]
        }

    fun info(message: String) {
        log(message, LEVEL.INFO)
    }

    private fun log(message: String, level: LEVEL) {
        findLogger(level).log(caller, Date().time, level, message, null)
    }

    fun log(message: String?, throwable: Throwable) {
        log(message, throwable, LEVEL.FATAL)
    }

    private fun log(message: String?, throwable: Throwable, level: LEVEL) {
        val logger = findLogger(level)
        logger.log(
            caller,
            Date().time,
            level,
            message ?: String.format("%s: %s", throwable.javaClass, throwable.localizedMessage),
            throwable
        )
    }

    fun warn(message: String) {
        log(message, LEVEL.WARN)
    }

    fun warn(msg: String?, throwable: Throwable) {
        log(msg, throwable, LEVEL.WARN)
    }
}
